### ISSUE TITLE
Pagination on index and tags pages

### ISSUE THREAD

**Issue by ahansson89** [ryanwiemer/gatsby-starter-gcn#25]
Would love to see an example on how to do this.

**Comment by ryanwiemer** [comment]
@ahansson89 great idea. To be honest I haven't tried it out yet but the plugin [gatsby-paginate](https://github.com/pixelstew/gatsby-paginate) looks useful. 

Feel free to give that a shot and submit a PR. Otherwise I may look into adding that functionality when updating the repo to use the `v2` version of Gatsby in the near future.

**Comment by ahansson89** [comment]
Appreciate it!

**Comment by ryanwiemer** [comment]
@ahansson89, 

I currently have a branch called [`paginate`](https://github.com/ryanwiemer/gatsby-starter-gcn/tree/paginate) to work on this functionality.

[Current Demo](https://paginate--gcn.netlify.com/)

Note that due to a limitation of the `gatsby-source-contentful` plugin I can't paginate the tag pages with how they are currently set up. To mitigate this I added a [load more button](https://paginate--gcn.netlify.com/tag/cute/).

I am open to suggestions if you or anyone else has any thoughts.

**Comment by ahansson89** [comment]
This is how I did it for tags:

```
const loadTags = new Promise((resolve, reject) => {
    graphql(`
      {
        allContentfulTag{
          edges {
            node {
              title
              id
              slug
              post {
                id
                title
                subTitle
                slug
                publishDate(formatString: "MMMM DD, YYYY")
                heroImage {
                  title
                  sizes(maxWidth: 1800) {
                    aspectRatio
                    src
                    srcSet
                    srcWebp
                    srcSetWebp
                    sizes
                  }
                }
                author {
                  authorName
                }
                tags {
                  slug
                  title
                  id
                }
                body {
                  childMarkdownRemark {
                    html
                    excerpt(pruneLength: 210)
                  }
                }
              }
            }
          }
        }
      }
    `).then(result => {
      result.data.allContentfulTag.edges.map(({ node }) => {
        createPaginatedPages({
          edges: sortBy(
            uniqBy(node.post || [], item => item.id),
            item => item.publishDate
          ).reverse(),
          createPage: createPage,
          pageTemplate: "./src/templates/tag.js",
          pageLength: config.postsPerPage,
          pathPrefix: `tag/${node.slug}`,
          context: {
            title: node.title,
            slug: node.slug
          }
        })
      })
      resolve()
    })
  })
```

Basically loading everything I need in gatsby-node.jd and passing that into the tag template like this:

```
const TagTemplate = ({ data, pathContext }) => {
  const { title, slug } = pathContext.additionalContext

  const { group, index, pageCount } = pathContext;
  const previousUrl = index - 1 == 1 ? "/" : "/" + (index - 1).toString() + "/" ;
  const nextUrl = "/" + (index + 1).toString() + "/" ;
  const indexSlug = index === 1 ? "" : index.toString() + "/"
  return (
    <div>
      <Helmet>
        <title>{`Tag: ${title} | ${config.siteTitle}`}</title>
        <meta
          property="og:title"
          content={`Tag: ${title} - ${config.siteTitle}`}
        />
        <meta property="og:url" content={`${config.siteUrl}/tag/${slug}/${indexSlug}`} />
      </Helmet>

      <div>
        <Heading.H1>Tag: {title}</Heading.H1>
        {group.map( post => (
          <PostListItem
            key={post.id}
            slug={post.slug}
            image={post.heroImage}
            title={post.title}
            subTitle={post.subTitle}
            date={post.publishDate}
            dateTime={post.publishDateISO}
            tags={post.tags}
            excerpt={post.body}
            author={post.author.authorName}
            avatar={post.author.avatar}
          />
        ))}
        <PaginationLinks index={index} pageCount={pageCount} previousUrl={previousUrl} nextUrl={nextUrl} />
      </div>
    </div>
  )
}
```

This code is not cleaned up yet as I was just focusing on getting it done quick.

**Comment by ryanwiemer** [comment]
@ahansson89,

Thanks for sharing some insight from your code! I was actually able to get the tags to paginate after all. The logic in my `gatsby-node.js` file is a little more complicated than I'd like but it works. I'll revisit this topic in the future and hopefully simplify it a bit more.

As a side note if I were to pick a plugin I like the philosophy of [gatsby-awesome-pagination](https://github.com/GatsbyCentral/gatsby-awesome-pagination#philosophy) over [gatsby-paginate](https://github.com/pixelstew/gatsby-paginate). I prefer keeping the data in the page queries. The only reason I am holding off on using a plugin now is due to needing additional features. i.e. Needing different number of posts on the first page.

**Comment by ryanwiemer** [comment]
Closed via https://github.com/ryanwiemer/gatsby-starter-gcn/pull/34 and released as [v2.0.0](https://github.com/ryanwiemer/gatsby-starter-gcn/releases/tag/v2.0.0)
